### PR TITLE
Replace primaryPostAllocated flag and use UnassignedInfo

### PR DESCRIPTION
**Pull request by kimchy**
There is no need to maintain additional state as to if a primary was allocated post api creation on the index routing table, we hold all this information already in the UnassignedInfo class.

### CODE REVIEW

File: core/src/main/java/org/elasticsearch/cluster/routing/allocation/command/AllocateAllocationCommand.java [hunk 1]
```diff
@@ -221,15 +221,17 @@ public RerouteExplanation execute(RoutingAllocation allocation, boolean explain)
...
                 continue;
             }
-            it.initialize(routingNode.nodeId());
-            if (shardRouting.primary()) {
-                // we need to clear the post allocation flag, since its an explicit allocation of the primary shard
-                // and we want to force allocate it (and create a new index for it)
-                routingNodes.addClearPostAllocationFlag(shardRouting.shardId());
+            // if we force allocation of a primary, we need to move the unassigned info back to treat it as if
+            // it was index creation
+            if (unassigned.primary() && unassigned.unassignedInfo().getReason() != UnassignedInfo.Reason.INDEX_CREATED) {
```
Comment (dakrone) [inline core/src/main/java/org/elasticsearch/cluster/routing/allocation/command/AllocateAllocationCommand.java]: Can `unassigned.unassignedInfo()` be null here? For instance, if someone tries to allocate an already allocated shard?

File: core/src/main/java/org/elasticsearch/cluster/routing/allocation/decider/DisableAllocationDecider.java [hunk 2]
```diff
@@ -109,7 +109,7 @@ public Decision canAllocate(ShardRouting shardRouting, RoutingNode node, Routing
             return allocation.decision(Decision.YES, NAME, "allocation disabling is ignored");
         }
         Settings indexSettings = allocation.routingNodes().metaData().index(shardRouting.index()).settings();
-        if (shardRouting.primary() && !allocation.routingNodes().routingTable().index(shardRouting.index()).shard(shardRouting.id()).primaryAllocatedPostApi()) {
+        if (shardRouting.primary() && shardRouting.allocatedPostIndexCreate() == false) {
```
Comment (dakrone) [inline core/src/main/java/org/elasticsearch/cluster/routing/allocation/decider/DisableAllocationDecider.java]: This is much cleaner and nicer to read!

File: core/src/main/java/org/elasticsearch/cluster/routing/ShardRouting.java [hunk 3]
```diff
@@ -431,6 +451,7 @@ void reinitializeShard() {
         version++;
         state = ShardRoutingState.INITIALIZING;
         allocationId = AllocationId.newInitializing();
+        this.unassignedInfo = new UnassignedInfo(UnassignedInfo.Reason.REINITIALIZED, null);
```
Comment (bleskes) [inline core/src/main/java/org/elasticsearch/cluster/routing/ShardRouting.java]: good catch ... I hope a test failed somewhere :)

File: core/src/main/java/org/elasticsearch/cluster/routing/allocation/command/AllocateAllocationCommand.java (hunk 1 above)
Comment (kimchy) [inline core/src/main/java/org/elasticsearch/cluster/routing/allocation/command/AllocateAllocationCommand.java]: we can't have a situation where the shard is unassigned and not have an UnassignedInfo (we assert for it on ShardRouting), and we iterate here over the unassigned shards so all should be well.

Comment (bleskes) [inline core/src/main/java/org/elasticsearch/cluster/routing/allocation/command/AllocateAllocationCommand.java]: we go through the unassigned shard list, so I think that's OK. right?

File: core/src/main/java/org/elasticsearch/cluster/routing/ShardRouting.java (hunk 3 above)
Comment (kimchy) [inline core/src/main/java/org/elasticsearch/cluster/routing/ShardRouting.java]: yep :)

**Comment by bleskes** [comment]
LGTM. A little dance for this flag going away...

File: core/src/main/java/org/elasticsearch/cluster/routing/allocation/command/AllocateAllocationCommand.java (hunk 1 above)
Comment (kimchy) [inline core/src/main/java/org/elasticsearch/cluster/routing/allocation/command/AllocateAllocationCommand.java]: yep

**Comment by dakrone** [comment]
LGTM also, left one question out of curiosity.

**Pull request merged**